feat: migrate to null-safety

arthurdenner opened the pull request:
Closes #6.

arthurdenner committed:
feat: update effective_dart to 1.3.0

arthurdenner committed:
feat: update mockito to 5.0.0 and Dart SDK to 2.12.0

arthurdenner committed:
feat: replace effective_dart with pedantic

arthurdenner committed:
refactor: address issues reported by pedantic

arthurdenner committed:
chore: update dependencies and Dart SDK for example

arthurdenner committed:
chore: update example project

arthurdenner committed:
feat: migrate library to null-safety

arthurdenner committed:
feat: migrate example to null-safety

arthurdenner committed:
test: setup GitHub Action for testing

arthurdenner committed:
chore: add --no-fatal-warnings flag to GitHub Action

kmcgill88 commented on example/lib/main.dart:
```diff
@@ -16,7 +16,7 @@ class _ExampleAppState extends State<ExampleApp> {
     return ThemeModeHandler(
       manager: ExampleThemeModeManager(),
-      builder: (ThemeMode themeMode) {
+      builder: (ThemeMode? themeMode) {
```
I don't think that you want the builder to pass a nullable `ThemeMode`.

kmcgill88 commented on lib/theme_mode_handler.dart:
```diff
@@ -6,7 +6,7 @@ import 'theme_mode_manager_interface.dart';
 class ThemeModeHandler extends StatefulWidget {
   /// Function that runs when themeMode changes.
-  final Widget Function(ThemeMode themeMode) builder;
+  final Widget Function(ThemeMode? themeMode) builder;
```
Again I don't think you want this builder to pass back a nullable here. I would guess that by default you get `ThemeMode.system`.

arthurdenner replied:
I thought that too but then realised that it may be nullable.

When it first loads, `themeMode` is `null` internally, then we load the theme using the `manager.loadThemeMode` and set it.
If `manager.loadThemeMode` returns an invalid value, we set it to `widget.defaultTheme` - (this piece of code](https://github.com/arthurdenner/theme_mode_handler/blob/feat/null-safety/lib/theme_mode_handler.dart#L60-L67)

But if `widget.withFallback` is `true` - the default behaviour - an empty `Container` it's returned - [this piece of code](https://github.com/arthurdenner/theme_mode_handler/blob/feat/null-safety/lib/theme_mode_handler.dart#L71-L73)
So, if `widget.withFallback` is `false`, then `widget.builder` _will_ be called with a `null` value.

arthurdenner replied:
Thanks for the feedback, @kmcgill88!

I thought that too but then realised that it may be `null`.

When it first loads, `themeMode` is `null` internally, then we load the theme using the `manager.loadThemeMode` and set it.
If `manager.loadThemeMode` returns an invalid value, we set it to `widget.defaultTheme` - [this piece of code](https://github.com/arthurdenner/theme_mode_handler/blob/feat/null-safety/lib/theme_mode_handler.dart#L60-L67)

But to prevent a "flash", if `widget.withFallback` is `true` - the default behaviour - an empty `Container` it's returned - [this piece of code](https://github.com/arthurdenner/theme_mode_handler/blob/feat/null-safety/lib/theme_mode_handler.dart#L71-L73)
So, if `widget.withFallback` is `false`, then `widget.builder` _will_ be called with a `null` value.

arthurdenner replied:
I tried to use the `late` keyword here but it'd mean that `themeMode` is not `null` - which isn't true in the last scenario I described above.

```dart
late ThemeMode _themeMode = widget.defaultTheme;
```

What are your thoughts?

kmcgill88 replied:
hmmm I see what you mean. Could you refactor [build](https://github.com/arthurdenner/theme_mode_handler/blob/feat/null-safety/lib/theme_mode_handler.dart#L70) to use a `FutureBuilder` with the Future from loading the theme from disk? You could remove `withFallback` and replace that with a `paceHolderWidget` to display while the theme loads. This way users can define their own behavior while the theme gets loaded from disk.

If you went this direction you could make the theme non-nullable b/c the builder wouldn't be called unless there was a theme. If no theme on disk, just use `defaultTheme` that you currently have.

kmcgill88 replied:
I think I'm biased to the theme passed via builder is non-nullable but then again users can easily take the builder result and do a default value themselves. 🤷🏻‍♂️ Getting the package to support null safety is the main thing. 😄

arthurdenner replied:
I really like your idea. It gives more control to the users too, which I prefer. I'll work on it. Thanks!

arthurdenner committed:
feat: remove nullability from themeMode

arthurdenner committed:
chore: improve example with placeholderWidget

arthurdenner committed:
chore: update tests (not fixed yet)

arthurdenner commented:
@kmcgill88, the PR is now updated with the approach you've suggested. I liked it but I've got a question:

`manager.loadThemeMode` may throw but I didn't add any handling for that. Any suggestions for a good approach? An `errorWidgetBuilder(snapshot.error)` maybe?

~~The tests will be updated soon, when I have a look at [mockito's null-safety guide](https://github.com/dart-lang/mockito/blob/master/NULL_SAFETY_README.md).~~

Tests are now updated and passing. I'm marking it as ready for review. 🎉

arthurdenner committed:
test: update tests to support null-safety

arthurdenner committed:
chore: inline linter rules from pedantic@1.11.0

deandreamatias commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,17 @@
+          java-version: '12.x'
+      - uses: subosito/flutter-action@v1
+        with:
+          flutter-version: '2.0.0'
```
You can remove this 2.0.0. with next line (channel: stable), get the last stable version

deandreamatias commented on analysis_options.yaml:
```diff
@@ -1 +1,58 @@
-include: package:effective_dart/analysis_options.yaml
```
Why remove effective dart? :(

arthurdenner replied:
Initially, I replaced it with `pedantic` because the GitHub Action was failing with it. Then I came across the same issue with `pedantic`.

It seems to be an issue in the `flutter analyze` command: it exits with `code 1` even when there are only warnings - [see the logs](https://github.com/arthurdenner/theme_mode_handler/runs/2045879652?check_suite_focus=true).

Since there isn't much to analyze, it was easier to replace the whole lint file with `pedantic@1.11.0`'s content directly.
If you know any workaround, I appreciate it. I tried `flutter analyze --no-fatal-warnings` too but it didn't work.

deandreamatias commented on pubspec.yaml:
```diff
@@ -4,15 +4,14 @@ version: 2.0.0
 homepage: 'https://github.com/arthurdenner/theme_mode_handler'
```
In line:2, you can change to `Flutter widget to change...` like a readme.

arthurdenner committed:
chore: remove hardcoded Flutter version and update package's description

arthurdenner commented:
@kmcgill88, pinging you about the latest changes in case you want to review them before I merge them.
